design/design-doc-integration-tests.md

```diff
@@ -0,0 +1,499 @@
+  resources to be tested for the same conditions.
+- It is pluggable, meaning that it can easily be incorporated into a new project
+  with minimal time and effort.
+- It is familiar, meaning using the framework should feel like writing other
+  tests in Go.
+- It is simple, meaning efforts to develop and maintain the framework should be
+  miniscule in comparison to the value it adds to the project.
+
+The testing framework should *not* achieve the following:
+
+- Replace the execution of unit tests.
+- Serve as a build / deployment system.
+
+The features above are already handled by robust platforms that are currently
+serving their purpose effectively. Addressing them in the initial implementation
+of this framework would require more effort than the return they would provide.
+
+## Definition of a Successful Test
+
+As mentioned above, the framework should unopinionated about the desired outcome
+of a test. Every test should be able to define what it deems to be successful
+and the framework should merely verify that that the successful conditions are
+met or not.
+
+### Definition of a Testable Unit
+
+Any controller that that is able to successfully register with the Kubernetes
+control plane should be testable. Any valid CustomResourceDefinitions that are
+able to be registered with the [Scheme] of the Kubernetes API should be
+testable.
+
+## Proposal
+
+The following sections describe a testing framework that utilizes the Go
+[testing] package to execute tests against any Kubernetes control plane. The
+framework is designed such that it can be minimally implemented in the
+short-term, deferring most of the execution and configuration to the tests that
+are written for each implementation. It is intended to evolve over time to
+reduce the amount of configuration required by each implementation, while still
+allowing flexibility to test any scenario.
+
+### System Design
+
+The framework should be made up of two core components:
+
+- `Job`: a testing environment. A job is made up of configuration, which informs
+  the framework of how to setup the control plane, and `Tests` which test actual
+  execution against the control plane.
+- `Test`: a test is an action or set of actions to execute against the
+  Kubernetes control plane. It could involve one or more steps, each of which
+  may cause the test to fail. Each test should be a logical operation, meaning
+  that tests that involve multiple steps should still be checking a single
+  logical condition (i.e. a `TestCreateSubnetSuccessful` may create the
+  `Subnet`, then wait to make sure that its `ConditionedStatus` is `Ready:
+  True`). 
+
+#### Jobs
+
+A `Job` contains the following components:
+
+```go
+// A Job is a set of tests that are executed sequentially in the same cluster environment
+type Job struct {
+    Name        string
+    Description string
+    Tests       []Test
+    cfg         *JobConfig
+    runner      *testing.T
```

Comment:
Could this be an argument of `Run` rather than a field of `Job`? I believe it's only used in `Run`.